torch_xla/csrc/aten_cpu_fallback.cpp

```diff
@@ -1,34 +1,362 @@
+}
+
+// Former 'to_cpu'.
+// In order to move tensors from XLA to CUDA, we make use of the DLPack API.
+//
+//   1. Synchronize the XLA tensors, so that we can access their data pointer
+//   2. Use DLPack in order to create a CUDA tensor
+static std::vector<at::Tensor> to_cuda(const at::TensorList& tensors,
+                                       at::DeviceIndex* common_device) {
+  // Synchronize tensors, so that we are able to grab their data pointer.
+  std::vector<XLATensorPtr> xla_tensors;
+  for (auto& tensor : tensors) {
+    if (is_valid_xla_tensor(tensor)) {
+      xla_tensors.push_back(get_xla_tensor(tensor));
+    }
+  }
+  XLAGraphExecutor::Get()->SyncTensorsGraph(
+      &xla_tensors, /*devices=*/{}, /*wait=*/true, /*sync_ltc_data=*/true,
+      /*warm_up_cache_only=*/false);
+
+  // Use DLPack for sharing the XLA storage with a newly created CUDA tensor.
+  std::vector<at::Tensor> cuda_tensors(tensors.size());
+  std::transform(tensors.begin(), tensors.end(), cuda_tensors.begin(),
+                 [=](const at::Tensor& tensor) {
+                   // Skip undefined or non-XLA tensors.
+                   if (!is_valid_xla_tensor(tensor)) {
+                     return tensor;
+                   }
+
+                   // Grab the DLManagedTensor.
+                   DLManagedTensor* managed = torch_xla::toDLPack(tensor);
+                   // Remember the device index, so that we can synchronize it
+                   // later. Also ensure that there's only one.
+                   at::DeviceIndex device = managed->dl_tensor.device.device_id;
+                   if (*common_device == -1) {
+                     *common_device = device;
+                   } else {
+                     TORCH_CHECK(*common_device == device,
+                                 "fallback using Multi-GPUs not supported");
+                   }
+                   // Create the CUDA tensor.
+                   return at::fromDLPack(
+                       managed, [=](void*) { managed->deleter(managed); });
+                 });
+  return cuda_tensors;
+}
+
+// Copy back the results from CUDA to XLA.
+// Assumes that we have already synchronized CUDA.
+static at::Tensor to_xla(const at::Tensor& tensor) {
+  return torch_xla::fromDLPack(at::toDLPack(tensor));
+}
+
+// Synchronizes the CUDA device being used by PyTorch.
+static void torch_cuda_synchronize(at::DeviceIndex common_device) {
+  // Save the current PyTorch device, in case it's not the same as the
+  // recorded tensor device.
+  at::DeviceIndex current = c10::cuda::current_device();
+  c10::cuda::set_device(common_device);
+  c10::cuda::device_synchronize();
+  c10::cuda::set_device(current);
+}
+
+// Former 'cpu_fallback'.
+// Changes:
+//
+//   1. Track the device index being used. Rationale: we synchronize the device
+//      before crossing device borders for correctness.
+//
+void cuda_fallback(const c10::OperatorHandle& op, torch::jit::Stack* stack,
```

Comment:
I found similar implementations in PyTorch, one in pytorch/aten/src/ATen/native/CPUFallback.cpp and the other one in pytorch/torch/csrc/lazy/ts_backend/ts_eager_fallback.cpp. How is this implementation different from those?

Reply:
The main difference is that we (a) are falling back to CUDA. Here are more details regarding these 3 implementations:

- _CPUFallback.cpp_ (b) looks more updated than _ts_eager_fallback.cpp_ (c), e.g. handles mutation in tensor lists
- **Device support:** even though (c) does support other devices, its conversion uses `tensor.to(device)`, which copies the tensor. In contrast, (a) we simply share the storage of the tensors
- **Device synchronization:** (a) needs further device synchronization, since we are not calling `tensor.to` method

With all that said, I do agree the functions are all similar to each other. A better approach would be to generalize (b), adding support for some customization.